### DISCUSSION TITLE
1. Introduction to JavaScript

### DISCUSSION THREAD

**Discussion by IamLizu** [IamLizu/jsland#23] [General]
**Disclaimer**: all materials are internet-sourced, nothing is originally written by me.

JavaScript is one of the most popular programming languages used to add interactive elements to web pages. It enables you to create dynamic content, control multimedia, and much more. Whether you're building web applications, mobile apps, or server-side programs, JavaScript plays a critical role in modern development.

* * *

### **What is JavaScript?**

JavaScript (JS) is a lightweight, interpreted, or just-in-time compiled programming language with first-class functions. It is one of the core technologies of the World Wide Web, alongside HTML and CSS.

-   **Client-Side**: Runs directly in the browser, providing interactivity (e.g., animations, form validations).
-   **Server-Side**: Using environments like Node.js, JavaScript can run on servers to handle backend logic.

### **Key Features of JavaScript**

-   **Interpreted language**: Runs directly in the browser.
-   **Event-Driven**: Executes code in response to events like user clicks or API responses.
-   **Weakly Typed**: Variable types are determined at runtime.
-   **Object-Oriented**: Supports object creation and manipulation.

* * *

### **Setting up Your Environment**

Before writing JavaScript code, let's set up the necessary tools:

1.  **Install Node.js**: Node.js allows you to run JavaScript outside the browser.
    
    -   Download Node.js and install it on your machine.
    -   To verify, open a terminal/command prompt and type:
        
        ```js
        node -v npm -v
        ```
        
    -   You should see the versions of Node.js and npm (Node Package Manager).
2.  **Install a Text Editor**:
    
    -   I recommend using [Visual Studio Code (VS Code)](https://code.visualstudio.com/).
    -   Install VS Code and open it to start writing JavaScript code.
3.  **Set Up a Basic Project**:
    
    -   Create a folder for your project.
    -   Inside your folder, create a file called `index.js`.
    -   Now you’re ready to write JavaScript!

* * *

### **Basic Syntax and Structure**

1.  **Variables and Data Types**
    
    Variables are containers for storing data values. In JavaScript, you can declare a variable using `var`, `let`, or `const`.
    
    ```js
    let name = "John";  // string
    const age = 30;     // number
    var isStudent = true; // boolean
    ```

    -   **Data Types**:
        -   Strings (`"Hello"`, `"JavaScript"`)
        -   Numbers (`10`, `4.5`)
        -   Booleans (`true`, `false`)
        -   Arrays (`[1, 2, 3]`)
        -   Objects (`{name: "John", age: 30}`)
2.  **Operators** JavaScript supports various operators:
    
    -   Arithmetic (`+`, `-`, `*`, `/`)
    -   Assignment (`=`, `+=`, `-=`)
    -   Comparison (`==`, `===`, `!=`, `<`, `>`)
    -   Logical (`&&`, `||`, `!`)
    
    Example:
    
    ```js
    let x = 5 + 10;  // 15
    let y = x * 2;   // 30
    ```
    
3.  **Control Structures** Control how your code runs:
    
    -   **if-else**: Conditional statements.
        
        ```js
        let score = 85;

        if (score >= 90) {
           console.log("Excellent");
        } else if (score >= 70) {
           console.log("Good");
        } else {
           console.log("Needs Improvement");
        }
        ```        
    -   **Loops**: Repeat code multiple times.
       
        ```js
        for (let i = 0; i < 5; i++) {
          console.log(i);
        }
        ```
        
4.  **Functions** Functions allow you to group code into reusable blocks.
    
    ```js
    function greet(name) {
      return "Hello, " + name;
    }
    console.log(greet("Alice"));
    ```
    
    -   Functions can also be anonymous (used in event handlers or callbacks).

* * *

## **Test Problems**

Here are a few beginner problems to practice your newly learned skills:

### 1\. **Problem 1: Variable Manipulation**

Write a program that:

-   Declares two variables `a` and `b` with values `5` and `10`.
-   Adds them together, stores the result in a new variable `sum`, and prints the result.
-   If the sum is greater than 10, print "Big Number", otherwise print "Small Number".

```js
// Expected Output: // 15 // Big Number
```

### 2\. **Problem 2: Even or Odd**

Write a function `isEven` that takes a number as input and returns:

-   `"Even"` if the number is divisible by 2.
-   `"Odd"` otherwise.

```js
// Example: isEven(4) should return "Even"
```

### 3\. **Problem 3: Array Operations**

Create an array of numbers. Write a program to:

-   Print the length of the array.
-   Add a new number to the end of the array.
-   Loop through the array and print each number.

```js
// Example: Given array [1, 2, 3], your program should:
// Print: 3 (length)
// Add 4 -> New array: [1, 2, 3, 4]
// Print: 1 2 3 4
```

### 4\. **Problem 4: Basic Calculator**

Create a simple calculator that prompts the user for two numbers and an operator (`+`, `-`, `*`, `/`). Perform the operation and print the result.

```js
// Example: If user enters 10, 5, and *, it should print: 50
```

* * *

Feel free to reach out if you have any questions, or if you get stuck on any of the problems! Happy learning! 🎉

Still hungry? Awesome (I hoped you would be hungry), read this: https://github.com/getify/You-Dont-Know-JS/blob/2nd-ed/get-started/README.md.

**Comment by Mehjabin08** [comment]
I'll set everything up as soon as possible. Some academic work is goin on. 
Will surely catch up in some days✊

**Reply by IamLizu** [reply]
Hey @Mehjabin08 

Any update on that?

**Comment by M-A-Mohit-Khan** [comment]
All the exercises are here- Hope to get feedback:
//1. Problem 1: Variable Manipulation
const num1= 5;
const num2= 10;

const sum= num1 + num2 ;
console.log(sum);
if(sum>10)
{
  console.log("Big Number");
}
else console.log("Small Number");

//2. Problem 2: Even or Odd
function isEven(number){
  if(number%2==0)
  {
    return "Even";
  }
  else return "Odd";
}
console.log(isEven(4));

//3. Problem 3: Array Operations
let arr=[1,2,3];
console.log(arr.length);
arr.push(4);
for(let i=0;i<arr.length;i++)
{
  console.log(arr[i]);
}

//4. Problem 4: Basic Calculator
let number1=parseInt(prompt("Input the 1st number"));
let operator=prompt("Input operators like +.-./.*");
let number2=parseInt(prompt("Input the 2nd number "));
if(operator=="+")
{
  console.log(number1+number2);
}
else if(operator=="-")
{
  console.log(number1-number2);
}
else if(operator=="/")
{
  console.log(number1/number2);
}
else  console.log(number1*number2);

**Reply by IamLizu** [reply]
Awesome @M-A-Mohit-Khan 

Please allow me to get back to you in a couple of days.

**Reply by M-A-Mohit-Khan** [reply]
Sure

**Comment by IamLizu** [comment]
We should have a short online meeting to discuss the above concepts. But only if you guys feel the need to. So, please let me know.

**Comment by IamLizu** [comment]
Hi 👋 

Now that we have a small filtered group of people, 

So I have decided to have a chat group for faster communication.

Please join https://m.me/cm/AbafOIJsJG0jBJmE/?send_source=cm:copy_invite_link

@M-A-Mohit-Khan @Mehjabin08 @parvezpappu
